src/frontend/src/expr/type_inference/func.rs

```diff
@@ -200,40 +234,114 @@ fn infer_type_for_special(
+                    input.cast_explicit(add_nesting(common_ele_type.clone(), x))
+                })
+                .try_collect();
+
+            match try_cast {
+                Ok(casted) => {
+                    // apply type conversion and return common type
+                    *inputs = casted;
+                    Ok(Some(array_type))
+                }
+                Err(_) => Err(ErrorCode::BindError(format!(
+                    "Cannot append {} to {}",
+                    right_type, left_type
+                ))
+                .into()),
+            }
         }
         ExprType::ArrayPrepend => {
             ensure_arity!("array_prepend", | inputs | == 2);
+            // get input types
             let left_type = inputs[0].return_type();
             let right_type = inputs[1].return_type();
-            let return_type = match (&left_type, &right_type) {
-                (DataType::List { .. }, DataType::List { .. }) => None,
-                (
-                    _, // non-array
-                    DataType::List {
-                        datatype: right_elem_type,
-                    },
-                ) if left_type == **right_elem_type => Some(right_type.clone()),
+            let right_ele_type_opt = match &right_type {
+                DataType::List { datatype: right_et } => Some(right_et),
                 _ => None,
             };
-            Ok(Some(return_type.ok_or_else(|| {
-                ErrorCode::BindError(format!("Cannot prepend {} to {}", left_type, right_type))
-            })?))
+            let right_ele_type = right_ele_type_opt.ok_or_else(|| {
+                ErrorCode::BindError(format!("Cannot prepend {} to {}", right_type, left_type))
+            })?;
+
+            // cast to least restrictive type or return error
+            let common_ele_type = least_restrictive(*right_ele_type.clone(), left_type.clone());
+            let nesting_level_diff = (calc_nesting_level(left_type.clone())
+                - calc_nesting_level(right_type.clone()))
+            .abs();
+            if common_ele_type.is_err() || nesting_level_diff > 1 {
+                return Err(ErrorCode::BindError(format!(
+                    "unable to find least restrictive type between {} and {}",
+                    left_type, right_type
+                ))
+                .into());
+            }
+
+            // found common type
+            let common_ele_type = common_ele_type.unwrap();
+            let most_nested = get_most_nested(left_type.clone(), right_type.clone());
+            let array_type = add_nesting(common_ele_type.clone(), most_nested);
+
+            // try to cast inputs to inputs to common type
+            let inputs_owned = std::mem::take(inputs);
+            let try_cast = inputs_owned
+                .into_iter()
+                .map(|input| {
+                    let x = input.return_type();
+                    input.cast_explicit(add_nesting(common_ele_type.clone(), x))
```

Comment:
Most automatically inserted casts should be `cast_implicit`, as suggested by the name. It is also the one used in `align_types` and `least_restrictive`. String concatenation is a very special case.